include/openenclave/internal/posix/sockops.h

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifndef _OE_POSIX_SOCKOPS_H
+#define _OE_POSIX_SOCKOPS_H
+
+#include <openenclave/bits/defs.h>
+#include <openenclave/bits/types.h>
+#include "deviceops.h"
+
+OE_EXTERNC_BEGIN
+
+typedef uint32_t socklen_t;
+struct oe_sockaddr;
+struct oe_addrinfo;
+struct oe_msghdr;
+
+typedef struct _oe_sock_ops
+{
+    oe_device_ops_t base;
+
+    oe_device_t* (
+        *socket)(oe_device_t* dev, int domain, int type, int protocol);
+
+    int (*connect)(
+        oe_device_t* dev,
+        const struct oe_sockaddr* addr,
+        socklen_t addrlen);
+
+    oe_device_t* (*accept)(
+        oe_device_t* dev,
+        struct oe_sockaddr* addr,
+        socklen_t* addrlen);
+
+    int (*bind)(
+        oe_device_t* dev,
+        const struct oe_sockaddr* addr,
+        socklen_t addrlen);
+
+    int (*listen)(oe_device_t* dev, int backlog);
+
+    ssize_t (*recv)(oe_device_t* dev, void* buf, size_t len, int flags);
+
+    ssize_t (*recvfrom)(
+        oe_device_t* dev,
+        void* buf,
+        size_t len,
+        int flags,
+        const struct oe_sockaddr* src_addr,
+        socklen_t* addrlen);
+
+    ssize_t (*send)(oe_device_t* dev, const void* buf, size_t len, int flags);
+
+    ssize_t (*sendto)(
+        oe_device_t* dev,
+        const void* buf,
+        size_t len,
+        int flags,
+        const struct oe_sockaddr* dest_addr,
+        socklen_t addrlen);
+
+    ssize_t (*socketpair)(
+        oe_device_t* dev,
+        int domain,
+        int type,
+        int protocol,
+        oe_device_t* retdevs[2]);
+
+    ssize_t (
+        *sendmsg)(oe_device_t* dev, const struct oe_msghdr* msg, int flags);
```

Comment:
Could you please help me by pointing me to
1. where oe_msghdr is serialized
2. How sendmsg gets translated into oe_sendmsg? I thought it would be under OE_NEEDSTDC_NAMES.

Reply:
This function pointer is called in posix/socket.c:

```
ssize_t oe_sendmsg(int sockfd, const struct oe_msghdr* buf, int flags)
{
    ssize_t ret = -1;
    oe_device_t* sock;

    if (!(sock = oe_fdtable_get(sockfd, OE_DEVICE_TYPE_SOCKET)))
        OE_RAISE_ERRNO(OE_EBADFD);

    ret = OE_CALL_SOCK(sendmsg, sock, buf, flags);

done:
    return ret;
}
```

The **oe_msghdr** structure is used throughout the enclave code. The code that serializes it and passes it to the host is defined by one of the devices (will be in the next PR).